Resume isolate before terminating tests to prevent flutter_tester leaks in integration tests

DanTup opened the pull request:
When running integration tests locally I noticed that `flutter_tester` processes were being leaked. This is because during some of the tests the Flutter isolate is left paused and when we try to tear down (by  sending `SIGTERM`) the output of the process says

> Waiting for current test to finish
> Press Ctrl+C again to terminate

Since the process doesn't quit within 10s, we then send `SIGTERM` which terminates the process and no cleanup is performed (so the child flutter_tester process is left). Sending a second `SIGTERM` did not resolve the issue in my testing because it also resulted in an unclean shutdown.

This probably fixes #26350, though I don't think it's a perfect fix since other code that terminates while paused (for example the user clicking Stop in the IDE while at a breakpoint) could still leak these processes - I don't think that's really a Flutter issue though (it's likely a `package:test` issue, or limitation that the IDEs should consider).

DanTup committed:
Resume isolate before terminating tests to prevent flutter_tester leak

DanTup committed:
Fix lint

DanTup committed:
Catch exceptions from resume() as well as timeout()

christopherfujino commented on packages/flutter_tools/test/integration.shard/test_driver.dart:
```diff
@@ -163,6 +163,14 @@ abstract class FlutterTestDriver {
+    // it forcefully and it won't terminate child processes, so we need to ensure
+    // it's running before terminating.
+    try {
+    await resume().timeout(defaultTimeout);
```
indentation?

zanderso reviewed (approved):
lgtm w/ indentation fix.

christopherfujino commented:
tools_tests are failing, can you try rebasing on ToT? this *might* have been fixed upstream.

DanTup committed:
Formatting fixes

DanTup commented:
Hmm, the failure does seem to be caused by this change - but I don't understand why the `try/catch` seems to be being ignored here and the error still thrown:

<img width="746" alt="Screenshot 2019-11-25 at 10 26 53 am" src="https://user-images.githubusercontent.com/1078012/69532646-2f251500-0f6e-11ea-9f2a-ee40691b4861.png">

The same code in a simple command line app seems to behave as I'd expect (the `catch` block executes). Is this a bug or am I overlooking something?

DanTup commented:
I've opened https://github.com/dart-lang/sdk/issues/39513 about the caught exception that's still failing the test (it's possible it's my bug, but I can't come up with an explanation that makes sense to me).

I'll see if I can find a better way to work around it here for now.

DanTup committed:
Don't try to resume if there's no vm service

DanTup commented:
I've rebased on latest code, fixed the indenting and also added a null check to avoid trying to resume when there's no VM service. The tests seem to all pass locally, though I'm still confused by the behaviour of the `catch` block - I may come back and tweak this in a future PR based on the response to https://github.com/dart-lang/sdk/issues/39513.

zanderso commented:
If `resume().timeout(defaultTimeout)` might leak an error into the containing Zone (via an unawaited `Future` or microtask), you can use [asyncGuard()](https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/base/async_guard.dart) to make sure it is caught where you want it to be caught.

However if `resume().timeout(defaultTimeout)` is in the tool codebase, the preferred solution is to not leak the error.

DanTup commented:
I was able to reproduce it without any zones in https://github.com/dart-lang/sdk/issues/39513 (and without the timeout firing). It feels like a bug, but I'm not sure 🤷‍♂️

zanderso commented:
I left a comment on https://github.com/dart-lang/sdk/issues/39513.

DanTup committed:
Fix handling of timeouts to not leak futures

DanTup commented:
I pushed 5207952f0 which changes the timeout code (which aimed to print a warning when an action took too long, including all of the messages that went over the API in the meantime if they're not already being printed) to use a `Timer` which is cancelled if the future completes.

This seems to resolve the issue - exceptions in `resume` are now handled and don't fail tests.

DanTup commented:
Looks like Cirrus builds aren't running for this PR again 😞

@fkorotkov - could be another instance of 404s from GH?

fkorotkov commented:
@DanTup there was another issue this time. I've poked around and found that GH timed out while delivering the webhook. Cirrus' metrics looks fine around that time so will investigate more to see on which end this happened. In the meantime I've manually redelivered the webhook.
